crates/matrix-sdk-base/src/rooms/normal.rs

```diff
@@ -1233,6 +1246,59 @@ impl Room {
         Ok(())
     }
 
+    /// Removes the seen knock request ids that are no longer valid given the
+    /// current room members.
+    pub async fn remove_outdated_seen_knock_requests_ids(&self) -> StoreResult<()> {
+        let not_knocked_memberships = RoomMemberships::all().difference(RoomMemberships::KNOCK);
+        let not_knocked_user_ids = BTreeSet::from_iter(
+            self.store.get_user_ids(self.room_id(), not_knocked_memberships).await?,
+        );
+
+        let mut current_seen_events_guard = self.seen_knock_request_ids_map.write().await;
+        // We're not calling `get_seen_join_request_ids` here because we need to keep
+        // the Mutex's guard until we've updated the data
+        let mut current_seen_events = if current_seen_events_guard.is_none() {
+            self.load_cached_knock_request_ids().await?
+        } else {
+            current_seen_events_guard.clone().unwrap()
+        };
+
+        // If there is nothing to compare the current seen knock ids to, return early
+        if not_knocked_user_ids.is_empty() {
+            return Ok(());
+        }
+
+        let mut to_remove = Vec::new();
+        for (event_id, user_id) in current_seen_events.iter() {
+            if not_knocked_user_ids.contains(user_id) {
+                to_remove.push(event_id.to_owned());
+            }
+        }
```

Comment:
I think the approach you've implemented works, but it does a lot of work when the room is large and when there's few members who knocked. Instead, could we iterate over all the user ids from the current_seen_events knock events, get their current membership from the store, and remove them if their own membership has changed?

Now that will mean that instead of loading N memberships (when N is the number of members in the room), we'll only load as many as the number of members who knocked.